examples/sui-move/hero/sources/example.move

```diff
@@ -0,0 +1,604 @@
+    const ENotAdmin: u64 = 3;
+
+    /// Not enough money to purchase the given item
+    const EInsufficientFunds: u64 = 5;
+
+    /// Trying to equip a sword but the hero already has one
+    const EAlreadyEquipped: u64 = 6;
+
+    /// Trying to remove a sword, but the hero does not have one
+    const ENotEquipped: u64 = 7;
+
+    // === Player Object creation ===
+
+    /// It all starts with the sword. Anyone can buy a sword, and proceeds are
+    /// stored in the `Game`. Amount of magic in the sword depends on how much
+    /// you pay for it.
+    public fun new_sword(
+        game: &mut Game,
+        payment: Coin<SUI>,
+        ctx: &mut TxContext
+    ): Sword {
+        let value = coin::value(&payment);
+        // ensure the user pays enough for the sword
+        assert!(value >= MIN_SWORD_COST, EInsufficientFunds);
+
+        // pay the game for this sword
+        balance::join(&mut game.payments, coin::into_balance(payment));
+
+        // magic of the sword is proportional to the amount you paid, up to
+        // a max. one can only imbue a sword with so much magic
+        let magic = (value - MIN_SWORD_COST) / MIN_SWORD_COST;
+        Sword {
+            id: object::new(ctx),
+            magic: math::min(magic, MAX_MAGIC),
+            strength: 1,
+            game_id: object::id(game)
+        }
+    }
+
+    /// Anyone can create a hero if they have a sword. All heroes start with the
+    /// same attributes.
+    public fun new_hero(sword: Sword, ctx: &mut TxContext): Hero {
+        Hero {
+            id: object::new(ctx),
+            game_id: sword.game_id,
+            health: 100,
+            experience: 0,
+            sword: option::some(sword),
+        }
+    }
+
+    // === Admin Object creation ===
+
+    /// Create a new `Game` (shared) and an `Admin` (returned) to run it. Anyone
+    /// can run a game, all objects spawned by the game will be associated with
+    /// it.
+    public fun new_game(ctx: &mut TxContext): Admin {
+        let game = Game {
+            id: object::new(ctx),
+            payments: balance::zero(),
+        };
+
+        let admin = Admin {
+            id: object::new(ctx),
+            game_id: object::id(&game),
+            boars_created: 0,
+            potions_created: 0,
+        };
+
+        transfer::share_object(game);
+        admin
+    }
+
+    /// Admin can create a potion with the given `potency` for `recipient`
+    public fun new_potion(
+        admin: &mut Admin,
+        potency: u64,
+        ctx: &mut TxContext
+    ): Potion {
+        admin.potions_created = admin.potions_created + 1;
+        Potion { id: object::new(ctx), potency, game_id: admin.game_id }
+    }
+
+    /// Admin can create a boar with the given attributes
+    public fun new_boar(
+        admin: &mut Admin,
+        health: u64,
+        strength: u64,
+        ctx: &mut TxContext
+    ): Boar {
+        admin.boars_created = admin.boars_created + 1;
```

Comment:
Love that some game state is carried with Admin.

Reply:
Yeah... it's a bit strange. I'm fairly sure this is just to demonstrate a richer kind of capability -- there's no reason to track these stats otherwise.